lib/AST/StmtManual.cpp

```diff
@@ -65,55 +65,166 @@ std::optional<Macro> Stmt::HighestContainingSubstitution(void) const noexcept {
-  //    expanded from and repeat steps 2 and 3.
-  //    If we reach a point where the first token is no longer expanded from a
-  //    macro argument, then return std::nullopt.
+  //    expanded from and repeat steps 2 and 3. If we reach a point where the
+  //    first token is no longer expanded from a macro argument, then return
+  //    std::nullopt.
+  
+  const auto last_tok = *(Tokens().end() - 1);
   for (auto begin = std::optional(*Tokens().begin()); begin;
        begin = begin->DerivedLocation()) {
-    auto macro_tok = begin->MacroLocation();
-    if (!macro_tok) {
-      continue;
+    if (auto macro_tok = begin->MacroLocation()) {
+      if (auto parent = macro_tok->Parent()) {
+        if (parent->Kind() == MacroKind::kArgument) {
+          if (last_tok.IsDerivedFromMacro(*parent)) {
+            return MacroArgument::From(parent.value());
+          }
+        }
+      }
     }
+  }
+  return std::nullopt;
+}
 
-    auto containing_macro = macro_tok->Parent();
-    if (containing_macro->Kind() == MacroKind::kArgument) {
-      if (std::all_of(
-        Tokens().begin(), Tokens().end(),
-        [&containing_macro](const Token token) {
-          return IsTokenDerivedFromMacro(token, containing_macro.value());
-        })) {
-        return MacroArgument::From(containing_macro.value());
+// Returns the lowest macro argument that covers this Stmt, if any.
+std::optional<MacroArgument>
+Stmt::LowestCoveringMacroArgument(void) const noexcept {
+  // Algorithm:
+  // 1. Walk down the Stmt's begin token's derivation tree until we reach a
+  //    token that was expanded from a macro argument, and is the first token
+  //    expanded from that argument. The Stmt aligns in the front with this
+  //    macro argument.
+  // 2. If we find such a macro argument, then walk down the Stmt's end token's
+  //    derivation tree until we reach a token that was expanded from a macro
+  //    argument, and is the last token expanded from that argument. The Stmt
+  //    aligns in the back with this macro argument. One extra wrinkle is that,
+  //    if while walking the end token's derivation tree we ever reach a point
+  //    where the begin token and end token are derived from the same token,
+  //    then we have to also check that the begin and end token are the same;
+  //    otherwise we must stop checking for back alignment at this point because
+  //    the begin and end tokens derivation trees have mixed and may fall under
+  //    the same macro argument, even though the begin and end tokens do not
+  //    actually expand from the same argument.
+  // 3. If we cannot find a back-aligned macro argument, then walk down the
+  //    begin token's derivation tree and try to find another front-aligned
+  //    macro argument.
+  // 4. If we cannot find a front-aligned macro argument, then return
+  //    std::nullopt.
+  Token stmt_begin_tok = BeginToken(), stmt_end_tok = EndToken();
+
```

Comment:
Worth checking `if (!stmt_begin_tok)`. Tokens are weird in that they can be nullish, because it was too much work when auto-generating the API to figure out what `clang::SourceLocation`s could plausible be invalid.